harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -636,3 +666,77 @@ def test_checkpoint_loading(ckpt: str, istrial: bool):
         assert isinstance(trial, pytorch.PyTorchTrial), type(trial)
     else:
         assert isinstance(trial, torch.nn.Module), type(trial)
+
+
+def make_amp_workloads(
+    trial_class,
+    assert_output_float16=False,
+) -> workload.Stream:
+    trainer = utils.TrainAndValidate()
+    yield from trainer.send(steps=20, validation_freq=1, scheduling_unit=1)
+    training_metrics, _ = trainer.result()
+
+    for metrics, next_metrics in zip(training_metrics[:-1], training_metrics[1:]):
+        scale = next_metrics["scale_before"]
+        if "scale" in metrics:
+            assert metrics["scale"] == scale
+        else:
+            metrics["scale"] = scale
+        assert "scale" in metrics
+    loss_prev = None
+    GROWTH_INTERVAL = trial_class._growth_interval
+    growth_countdown = GROWTH_INTERVAL
+    # Only attempt assertions up to and including the penultimate batch, because
+    #  we may not have the updated scale from the final batch.
+    for idx, metrics in enumerate(training_metrics[:-1]):
+        assert metrics["loss"].dtype is np.dtype("float32")
+        if assert_output_float16:
+            assert metrics["output"].dtype is np.dtype("float16"), (
+                f"output has dtype {metrics['output'].dtype} but should be float16"
+            )
+        else:
+            # Automatic usages of Apex or native AMP cast the output back to float32
+            # For the latter case, see the hook end_f16
+            #   defined in PyTorchTrialContext.autocast_forward_pass
+            assert metrics["output"].dtype is np.dtype("float32"), (
+                f"output has dtype {metrics['output'].dtype} but should be float32"
+            )
+        loss = metrics["loss"].item()
+        scale_before = metrics["scale_before"]
+        scaled_loss = loss * scale_before
+        if "scaled_loss" in metrics:
+            assert scaled_loss == metrics["scaled_loss"]
+        growth_countdown -= 1
+        scale = metrics["scale"]
+        MIN_SCALED_LOSS_TO_REDUCE_SCALE = 32760
+        if scale != scale_before:
+            if scale < scale_before:
+                assert scaled_loss >= MIN_SCALED_LOSS_TO_REDUCE_SCALE, (
+                    f"scale reduced from {scale_before} to {scale}, "
+                    f"but not as expected (scaled_loss={scaled_loss} < "
+                    f"{MIN_SCALED_LOSS_TO_REDUCE_SCALE})"
+                )
+            else:
+                assert growth_countdown == 0, (
+                    f"scale grew from {scale_before} to {scale}, "
+                    f"but not when expected (growth_countdown={growth_countdown:d})"
+                )
+            growth_countdown = GROWTH_INTERVAL
+        else:
+            assert (
+                growth_countdown != 0
+            ), f"scale was expected to grow (growth_countdown={growth_countdown}) but did not"
```

Comment:
The logic here seems backwards from what I would expect.

I would expect something logic to flow from what is known to what is expected:
```python
if growth_countdown == 0:
    assert scale > scale_before
else:
    if scaled_loss >= MIN_SCALED_LOSS_TO_REDUCE_SCALE:
        assert scale < scale_before
    else:
        ...
```

(not saying that my pseudo code is what you should write, but that it looks at known quantities to figure out what assertions to make, rather than looking at results and making assertions about known quantities.)

To me, that's part of the pattern of "calculate a ground truth from first principles and compare results to the ground truth".

Reply:
Good call! This is done.